src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+#include "foundation/math/distance.h"
+#include "foundation/math/knn/knn_statistics.h"
+#include "foundation/math/knn/knn_tree.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+
+// Standard headers.
+#include <cassert>
+#include <cstddef>
+#include <limits>
+
+// Enable or disable k-nn query statistics.
+#undef FOUNDATION_KNN_ENABLE_QUERY_STATS
+
+namespace foundation {
+namespace knn {
+
+template <typename T, size_t N>
+class AnyQuery
+  : public NonCopyable
+{
+  public:
+    typedef T ValueType;
+    static const size_t Dimension = N;
+
+    typedef Vector<T, N> VectorType;
+    typedef Tree<T, N> TreeType;
+
+    explicit AnyQuery(const TreeType& tree);
+
+    bool run(
+        const VectorType&   query_point,
+        const ValueType     query_max_square_distance
+#ifdef FOUNDATION_KNN_ENABLE_QUERY_STATS
+        , QueryStatistics&  stats
+#endif
+        ) const;
+
+#ifdef FOUNDATION_KNN_ENABLE_QUERY_STATS
+    bool run(
+        const VectorType&   query_point,
+        const ValueType     query_max_square_distance) const;
+#endif
+
+  private:
+    typedef typename TreeType::NodeType NodeType;
+
+    struct NodeEntry
+    {
+        ValueType           m_dvec_square_norm;
+        const NodeType*     m_node;
+        VectorType          m_dvec;
+
+        NodeEntry() {}
+
+        NodeEntry(
+            const NodeType*     node,
+            const VectorType&   dvec,
+            const ValueType     dvec_square_norm)
+          : m_dvec_square_norm(dvec_square_norm)
+          , m_node(node)
+          , m_dvec(dvec)
+        {
+        }
+
+        bool operator<(const NodeEntry& rhs) const
+        {
+            return m_dvec_square_norm > rhs.m_dvec_square_norm;
+        }
+    };
+
+    const TreeType& m_tree;
+};
+
+typedef AnyQuery<float, 2>  AnyQuery2f;
+typedef AnyQuery<double, 2> AnyQuery2d;
+typedef AnyQuery<float, 3>  AnyQuery3f;
+typedef AnyQuery<double, 3> AnyQuery3d;
+
+
+//
+// Implementation.
+//
+
+#ifdef FOUNDATION_KNN_ENABLE_QUERY_STATS
+#define FOUNDATION_KNN_QUERY_STATS(x) x
+#else
+#define FOUNDATION_KNN_QUERY_STATS(x)
+#endif
+
+template <typename T, size_t N>
+inline AnyQuery<T, N>::AnyQuery(const TreeType& tree)
+  : m_tree(tree)
+{
+}
+
+template <typename T, size_t N>
+inline bool AnyQuery<T, N>::run(
+    const VectorType&       query_point,
+    const ValueType         query_max_square_distance
+#ifdef FOUNDATION_KNN_ENABLE_QUERY_STATS
+    , QueryStatistics&      stats
+#endif
+    ) const
+{
+    assert(!m_tree.empty());
+
+    FOUNDATION_KNN_QUERY_STATS(++stats.m_query_count);
+    FOUNDATION_KNN_QUERY_STATS(size_t fetched_node_count = 0);
+    FOUNDATION_KNN_QUERY_STATS(size_t visited_leaf_count = 0);
+    FOUNDATION_KNN_QUERY_STATS(size_t tested_point_count = 0);
+
+    const VectorType* APPLESEED_RESTRICT points = &m_tree.m_points.front();
+    const NodeType* APPLESEED_RESTRICT nodes = &m_tree.m_nodes.front();
```

Comment:
> What is the benefit of having a pointer to a reference?

There's no such thing as a pointer to a reference in C++. A reference has no existence of its own (unlike a pointer which is a real object that takes space). Here we're simply taking a pointer to the front node.

> Also, why not make nodes a const std::vector<NodeType>&?

That would work but it would mean using `&nodes[0]` whenever we need to construct a pointer to one of the child node. It's likely that it would lead to the same generated code. It's a style issue I guess?

> If you want to stick to the pointer to a reference, I'd recommend renaming nodes to root, since this is, what front() actually returns.

`root` would be a good name too, maybe even better than `nodes`. The reason why it's called `nodes` is because it's an array of nodes (and you can do things like `nodes[nodes_index]` to reach a particular node), at least that's how I thought of it when writing this code.

Reply:
> It's a style issue I guess?
Yes, most of the other comments are style/readability issues. Readabilitywise I'd prefer the `&nodes[0]`, but then again, it might not be worth the effort of changing everything here.

> it's an array of nodes
Totally makes sense. I just wanted to raise another point of view on it. If you decide to stick to the current version, that's fine as well.

Reply:
I've renamed `nodes` to `root_node`.